### PR TITLE
Llama 3 Tokenizer Fixes, Cleanup of `resources.go`, upgrade of Go to 1.22.

### PR DESCRIPTION
**Pull request by wbrown**
* Implement `SplitRegex` 
LLama 3 has its own split regex that is slightly modified from the standard one, that caps numerics to three digits. As it is a backtracking regex, we have to monkey-patch whatever regex is passed in.

* Implement `IgnoreMerges`
New to Llama 3 is a new flag `IgnoreMerges`, where if a `word` passed into the `ToBPE` function can be directly looked up in the token table, we don't split it.

* Cleanup of `resources.go`
There are a lot of functions that operate on `Resources` structures, so I cleaned it up so that they are all receiver methods. There were also a lot of repeated code, so I implemented `UnmarshalData`, and `UnmarshalUntilData` to DRY things up.

* Upgraded to go 1.22.

### CODE REVIEW

**Review by Rexwang8 — APPROVED** [review]
looks good! no issues stand out to me

File: cmd/dataset_tokenizer/go.mod [hunk 1]
```diff
@@ -1,6 +1,6 @@
 module github.com/wbrown/gpt_bpe/cmd/dataset_tokenizer
 
-go 1.18
+go 1.22
```
Comment (Rexwang8) [inline cmd/dataset_tokenizer/go.mod:3]: Have you confirmed that all the commands, tests and other affected components of gpt_bpe work during the shift from 1.18 to 1.22?

Comment (wbrown) [reply cmd/dataset_tokenizer/go.mod:3]: All test cases passed. This was prompted by the debugger complaining about depreciation of `go 1.18`.

File: resources/resolver.go [hunk 2]
```diff
@@ -637,6 +637,41 @@ func ResolveResources(
...
+			if pretokenizers, ok := preTokMap["pretokenizers"]; ok &&
+				pretokenizers != nil {
+				pretokenizersList := pretokenizers.([]interface{})
+				for _, v := range pretokenizersList {
+					vIntf := v.(map[string]interface{})
+					if vIntf["type"] == "Split" {
+						pattern := vIntf["pattern"].(map[string]interface{})
+						if pattern["Regex"] != nil {
+							splitRegexVal := pattern["Regex"].(string)
+							// Fix lookbacks
```
Comment (dmarx) [inline resources/resolver.go:660]: <migrated offline clarification discussion>
me: The "fix" here is a performance improvement?
wes: No. It's a negative lookback. And golang's regex doesn't support it.

**Review by rtalaricw — APPROVED** [review]
Few comments:

- Naming conventions have really improved
- DRY to remove repeated code

We should always write code with readable and modular behavior where each unit of functionality can be tested independently. 

I do not have any other comments and approved it. Sorry for the delay.

**Pull request merged**